// vite-plugin-zip-pack/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const jszip_1 = __importDefault(require("jszip"));
function timeZoneOffset(date) {
    return new Date(date.getTime() - date.getTimezoneOffset() * 60000);
}
function zipPack(options) {
    const inDir = (options === null || options === void 0 ? void 0 : options.inDir) || "dist";
    const outDir = (options === null || options === void 0 ? void 0 : options.outDir) || "dist-zip";
    const outFileName = (options === null || options === void 0 ? void 0 : options.outFileName) || "dist.zip";
    const pathPrefix = (options === null || options === void 0 ? void 0 : options.pathPrefix) || '';
    function addFilesToZipArchive(zip, inDir) {
        const listOfFiles = fs_1.default.readdirSync(inDir);
        listOfFiles.forEach((fileName) => {
            const filePath = path_1.default.join(inDir, fileName);
            const file = fs_1.default.statSync(filePath);
            const timeZoneOffsetDate = timeZoneOffset(new Date(file.mtime));
            if (file === null || file === void 0 ? void 0 : file.isDirectory()) {
                zip.file(fileName, null, {
                    dir: true,
                    date: timeZoneOffsetDate
                });
                const dir = zip.folder(fileName);
                addFilesToZipArchive(dir, filePath);
            }
            else {
                zip.file(fileName, fs_1.default.readFileSync(filePath), { date: timeZoneOffsetDate });
            }
        });
    }
    function createZipArchive(zip) {
        // @ts-ignore
        zip.root = '';
        zip
            .generateAsync({
            type: "nodebuffer",
            compression: "DEFLATE",
            compressionOptions: {
                level: 9,
            },
        })
            .then((file) => {
            const fileName = path_1.default.join(outDir, outFileName);
            if (fs_1.default.existsSync(fileName)) {
                fs_1.default.unlinkSync(fileName);
            }
            fs_1.default.writeFileSync(fileName, file);
        });
    }
    return {
        name: "vite-plugin-zip-pack",
        apply: "build",
        closeBundle() {
            try {
                console.log("\x1b[36m%s\x1b[0m", `Zip packing - "${inDir}" folder :`);
                if (fs_1.default.existsSync(inDir)) {
                    if (!fs_1.default.existsSync(outDir)) {
                        fs_1.default.mkdirSync(outDir);
                    }
                    if (pathPrefix && path_1.default.isAbsolute(pathPrefix)) {
                        throw new Error('"pathPrefix" must be a relative path');
                    }
                    const zip = new jszip_1.default();
                    let archive;
                    if (pathPrefix) {
                        const timeZoneOffsetDate = timeZoneOffset(new Date());
                        zip.file(pathPrefix, null, {
                            dir: true,
                            date: timeZoneOffsetDate
                        });
                        archive = zip.folder(pathPrefix);
                    }
                    else {
                        archive = zip;
                    }
                    console.log("\x1b[32m%s\x1b[0m", "  - Preparing files.");
                    addFilesToZipArchive(archive, inDir);
                    console.log("\x1b[32m%s\x1b[0m", "  - Creating zip archive.");
                    createZipArchive(archive);
                    console.log("\x1b[32m%s\x1b[0m", "  - Done.");
                }
                else {
                    console.log("\x1b[31m%s\x1b[0m", `  - "${inDir}" folder does not exist!`);
                }
            }
            catch (error) {
                if (error) {
                    console.log("\x1b[31m%s\x1b[0m", `  - ${error}`);
                }
                console.log("\x1b[31m%s\x1b[0m", "  - Something went wrong while building zip file!");
            }
        },
    };
}
exports.default = zipPack;
